log.hpp

```diff
@@ -0,0 +1,54 @@
+
+#pragma once
+
+#include <utility>
+#include <cstdio>
+#include <cstdlib>
+
+inline void single_print(const char * arg)
+{
+    std::printf("%s ", arg);
+}
+
+inline void single_print(char arg)
+{
+    std::printf("%c ", arg);
+}
+
+inline void single_print(int arg)
+{
+    std::printf("%d ", arg);
+}
+
+inline void single_print(float arg)
+{
+    std::printf("%f ", arg);
+}
+
+inline void single_print(double arg)
+{
+    std::printf("%lf ", arg);
+}
+
+inline void single_print(bool arg)
+{
+    std::printf("%s ", arg ? "true" : "false");
+}
+
+template <typename Last>
+inline void print(Last last)
+{
+    single_print(last); std::printf("\n");
+}
+
+template <typename First, typename Second, typename ...Rest>
+inline void print(First first, Second second, Rest... rest)
+{
+    single_print(first); print(second, rest...);
+}
```

Comment:
You can just write:
```cpp
template <typename First, typename ...Rest>
inline void print(First first, Rest... rest) {
  print(first);
  print(rest);
}
```
and replacing `single_print()` with `print()`.

Reply:
Yeah if the newline printing is going to get separated from the rest of the function this is understandable.

Reply:
Even if it wasn't, it's not worth it. That's a personal opinion of course and in performance circumastances that I've not seen that might matter. But generally, I think that in big projects it's a bad idea to sacrifice so much readability for such a little perf benefit.

Reply:
Of course I understand. What I meant was that if you got just a single separator it makes things much simpler.

Reply:
No, AFAIK (and I don't pretend to be a C++ guru, neither would I ever waste time to become one) this is function overloading. Template specialization is what you did before.